js/iframe-coordinator/libs/workers/worker-manager.ts

```diff
@@ -0,0 +1,288 @@
+  * stronger types of messages between this manager and the workers?
+ * docs
+ *   dont forget that this is sorta dangerous with indexeddb
+ *
+ * Is the import of the worker good enough?
+  * Decide on class or text of the worker
+ * Make sure es6 target is ok
+ * Make sure it works in IE
+ *  check blob and fallback
+ * Improve worker demo for the real, useful demo.
+ *  use subscribe?
+ *  allow dynamic addition?
+ * Tests
+ *  tests for error rate
+ * better logging (these are important)
+ * support lenient shutdown?
+ *  notification on shutdown?
+ *  soft stop
+ *  allow restart?
+ * support client stop request?
+ *   worker never can close now as i have attached listeners
+ *   but, you can't re-add now either via registerclients
+ */
+
+enum WorkerPhase {
+  LOADING,
+  LOADED,
+  BOOTSTRAPPING,
+  RUNNING
+  // TODO Might want to implement this.  give it a chance to shutdown.  Need a timeout
+  // STOPPING
+}
+
+interface ManagedWorker {
+  id: string;
+  url: string;
+  phase: WorkerPhase;
+  worker: Worker;
+  errorWindowCount: number;
+  errorWindowStartTimestamp: number;
+}
+
+export default class WorkerManager implements EventListenerObject, EventTarget {
+  private static trackedWorkerEventTypes = ['message', 'error'];
+  private static evtTargetDelegate: DocumentFragment = null;
+  private static ID_PREFIX = 'iframeCoordinatorWorker-';
+  private static _workerIndex = 0;
+  private _workers: ManagedWorker[];
+  private _errorWindowCountThreshold = 0;
+  private _errorWindowMillis = 0;
+
+  constructor(
+    errorWindowCountThreshold: number = DEFAULT_ERROR_WINDOW_COUNT_THRESHOLD,
+    errorWindowMillis: number = DEFAULT_ERROR_WINDOW_MILLIS
+  ) {
+    this._errorWindowCountThreshold = errorWindowCountThreshold;
+    this._errorWindowMillis = errorWindowMillis;
+
+    if (WorkerManager.evtTargetDelegate === null) {
+      WorkerManager.evtTargetDelegate = document.createDocumentFragment();
+    }
+
+    this._workers = [];
+  }
+
+  public start(url: string): string {
+    const id = `${WorkerManager.ID_PREFIX}${++WorkerManager._workerIndex}`;
+
+    // @ts-ignore
+    const worker = new SpawnWorker();
+    WorkerManager.trackedWorkerEventTypes.forEach(curr => {
+      worker.addEventListener(curr, this);
+    });
+
+    this._workers.push({
+      id,
+      url,
+      worker,
+      phase: WorkerPhase.LOADING,
+      errorWindowCount: 0,
+      errorWindowStartTimestamp: -1
+    });
+
+    return id;
+  }
+
+  public stop(idToStop: string) {
+    const workerIndexToStop = this._workers.findIndex((curr, index) => {
+      return curr.id === idToStop;
+    });
+
+    if (workerIndexToStop >= 0) {
+      const worker = this._workers[workerIndexToStop].worker;
+      WorkerManager.trackedWorkerEventTypes.forEach(curr => {
+        worker.removeEventListener(curr, this);
+      });
+      worker.terminate();
+
+      this._workers.splice(workerIndexToStop, 1);
+    }
+  }
+
+  // TODO Need comms to workers (pub/sub, others?)
```

Comment:
@MattCheely Should background workers be able to subscribe/publish to any channel?  Should the host app be able to "authorize" or intercept and `preventDefault`?